### ISSUE TITLE
[YSQL] prepared statement with bigint paramType on integer column table takes significantly more time compared to postgres

### ISSUE THREAD

**Issue by shaharuk-yb** [yugabyte/yugabyte-db#12974]
Jira Link: [DB-2703](https://yugabyte.atlassian.net/browse/DB-2703)
### Description

Issue: When table is created with integer type column and prepared statement provides BIGINT as paramType, yugabyte's translateText takes more time compared to plain postgres.
It takes >=1 sec to process the same range query with BIGINT data type and takes few milliseconds to run same query with INT data type **through prepared statements.**

Use following C program to reproduce the issue.
to run:
```
gcc ps_test.c -lpq -I<yugabyte-install-dir>/postgres/include -o pstest
./pstest
```

```
#include <stdio.h>
#include <stdlib.h>
#include "libpq-fe.h"
#include <time.h>

typedef enum
{
  DB_TYPE_INT,
  DB_TYPE_BIGINT
} db_bind_type_t;

typedef struct
{
  db_bind_type_t   db_type;
  int              pg_type;
} db_pgsql_bind_map_t;

/* DB-to-PgSQL bind types map */
db_pgsql_bind_map_t db_pgsql_bind_map[] =
    {
        {DB_TYPE_INT,       23},
        {DB_TYPE_BIGINT,    20}
    };


int main()
{
  /* connection string */
  const char *conninfo = "host=127.0.0.1 port=5433 dbname=yugabyte user=yugabyte password=";

  /* Make a connection to the database */
  PGconn *conn = PQconnectdb(conninfo);
  /* Check to see that the backend connection was successfully made */
  if (PQstatus(conn) != CONNECTION_OK)
  {
    fprintf(stderr, "Connection to database failed: %s",
            PQerrorMessage(conn));
    PQfinish(conn);
    exit(1);
  }
  /* Use following sysbench commands to cleanup, create and load data
   * --serial_cache_size=1000 --table_size=100000 --tables=10 --thread-init-timeout=30 --time=30 --warmup-time=0 --create_secondary=true --range_key_partitioning=true --range_selects=true --db-driver=pgsql --pgsql-db=yugabyte --pgsql-port=5433 --pgsql-user=yugabyte  --pgsql-host=127.0.0.1 --threads=1 --skip_trx=true --range_size=100 cleanup
   * --serial_cache_size=1000 --table_size=100000 --tables=10 --thread-init-timeout=30 --time=30 --warmup-time=0 --create_secondary=true --range_key_partitioning=true --range_selects=true --db-driver=pgsql --pgsql-db=yugabyte --pgsql-port=5433 --pgsql-user=yugabyte  --pgsql-host=127.0.0.1 --threads=1 --skip_trx=true --range_size=100 create
   * --serial_cache_size=1000 --table_size=100000 --tables=10 --thread-init-timeout=30 --time=30 --warmup-time=0 --create_secondary=true --range_key_partitioning=true --range_selects=true --db-driver=pgsql --pgsql-db=yugabyte --pgsql-port=5433 --pgsql-user=yugabyte  --pgsql-host=127.0.0.1 --threads=1 --skip_trx=true --range_size=100 load
   * */

  /*
   * The Tables are created with integer data type for id field.
   * Now lets create prepared statements with int paramTypes
   * */
  const Oid oid[] = {(const Oid) db_pgsql_bind_map[DB_TYPE_INT].pg_type, (const Oid) db_pgsql_bind_map[DB_TYPE_INT].pg_type};

  PQprepare(conn, "sshaikh-1", "SELECT c FROM sbtest1 WHERE id BETWEEN $1 AND $2",
            2, oid);

  time_t start_t, end_t;
  const char *a[] = {"39652", "39751"};
  time(&start_t);
  PGresult  *res = PQexecPrepared(conn, "sshaikh-1", 2, a,
                                  (const int *) 2, NULL, 1);

  time(&end_t);
  PQclear(res);
  printf("Execution time with int paramType= %f\n", difftime(end_t, start_t));


  /*With bigint type paramTypes*/
  const Oid oid1[] = {(const Oid) db_pgsql_bind_map[DB_TYPE_BIGINT].pg_type, (const Oid) db_pgsql_bind_map[DB_TYPE_BIGINT].pg_type};
  PQprepare(conn, "sshaikh-2", "SELECT c FROM sbtest1 WHERE id BETWEEN $1 AND $2",
            2, oid1);

  const char *a1[] = {"39652", "39751"};
  time(&start_t);
  PGresult  *res1 = PQexecPrepared(conn, "sshaikh-2", 2, a1,
                                  (const int *) 2, NULL, 1);

  time(&end_t);
  PQclear(res1);
  printf("Execution time with BIGINT paramType= %f\n", difftime(end_t, start_t));
  PQfinish(conn);
  return 0;
}
```

Output:
```
Execution time with int paramType= 0.000000
Execution time with BIGINT paramType= 1.000000

```

**Comment by fritshoogland-yugabyte** [comment]
Use [yb_stats](https://github.com/fritshoogland-yugabyte/yb_stats.git) to analyze the time:

Well performing case:

1. YSQL statements
```
1 avg:           2.507 tot:           2.507 ms avg:          1 tot:          1 rows: explain analyze execute q_int(39652, 39751)
```
(1 execution, avg&total time of 2.5 ms, 1 row returned)
2. YSQL statistics
```
handler_latency_yb_ysqlserver_SQLProcessor_SelectStmt 1 avg: 1.773 tot:           1.773 ms, avg:             100 tot:             100 rows
```
(There were 1 select statements, avg&total time of 1.78 ms and it returned 100 rows)
3. tablet server: PgClientService_Perform
```
handler_latency_yb_tserver_PgClientService_Perform    1   0.119 /s avg:      1148 tot:            1148 ?
```
1 YSQL <> tserver request performed , which took 1.15ms.
4. tablet server: TabletServerService_Read
```
handler_latency_yb_tserver_TabletServerService_Read   1   0.119 /s avg:       976 tot:             976 us
```
1 tserver read request performed, which took 0.98ms.

What this means is that the execution of this prepared statement took 2.51ms, of which 1.15ms was spent in tserver.

Bad performing case:
1. YSQL statements
```
1 avg:         274.314 tot:         274.314 ms avg:          1 tot:          1 rows: explain analyze execute q_bigint(39652, 39751)
```
(1 execution, avg&total time 274.31ms, 1 row returned)
2. YSQL statistics
```
handler_latency_yb_ysqlserver_SQLProcessor_SelectStmt 1 avg:         273.366 tot:         273.366 ms, avg:             100 tot:             100 rows
```
(There were 1 select statements, avg&total time of 273.37 ms and it returned 100 rows)
3. tablet server: PgClientService_Perform
```
handler_latency_yb_tserver_PgClientService_Perform    49    8.453 /s avg:      4945 tot:          242308 ?
```
49 YSQL <> tserver requests performed, avg. response time 4.95ms, total time 242.31 ms.
4. tablet server: TabletServerService_Read
```
handler_latency_yb_tserver_TabletServerService_Read   49    8.453 /s avg:      4864 tot:          238336 us
```
49 read requests performed, avg. response time 4.87ms, total time 238.34 ms.

What this means is that the execution of this prepared statement took 274.31 ms, of which 242.31 ms was spent in tserver.

In other words: the reason the bad performing statement is taking its time, is because it spends most of it in tserver.

**Comment by fritshoogland-yugabyte** [comment]
The reason it is spending so much time in tserver in the bad performing case:

These are the rocksdb statistics of the well performing SQL:
```
tablet   rocksdb_db_iter_bytes_read           51736 bytes         6164.919 /s
tablet   rocksdb_no_table_cache_iterators         1 iters            0.119 /s
tablet   rocksdb_number_db_next                 400 keys            47.664 /s
tablet   rocksdb_number_db_next_found           400 keys            47.664 /s
tablet   rocksdb_number_db_seek                 201 keys            23.951 /s
tablet   rocksdb_number_db_seek_found           201 keys            23.951 /s
```
And these are the rocksdb statistics of the bad performing SQL:
```
tablet   rocksdb_db_iter_bytes_read                  25820255 bytes      4454071.934 /s
tablet   rocksdb_no_table_cache_iterators                  49 iters            8.453 /s
tablet   rocksdb_number_db_next                        200191 keys         34533.552 /s
tablet   rocksdb_number_db_next_found                  200191 keys         34533.552 /s
tablet   rocksdb_number_db_seek                        100145 keys         17275.315 /s
tablet   rocksdb_number_db_seek_found                  100144 keys         17275.142 /s
```

These are the rocksdb statistics of a seq scan: (`explain analyze select c from performance`):
```
tablet   rocksdb_db_iter_bytes_read       25820255 bytes      1499782.470 /s
tablet   rocksdb_no_table_cache_iterators       49 iters            2.846 /s
tablet   rocksdb_number_db_next             200191 keys         11628.195 /s
tablet   rocksdb_number_db_next_found       200191 keys         11628.195 /s
tablet   rocksdb_number_db_seek             100145 keys          5816.973 /s
tablet   rocksdb_number_db_seek_found       100144 keys          5816.914 /s
```

In other words: in the bad performing SQL, despite showing itself as performing an index scan, the whole PK, and therefore the whole table is read, with identical rocksdb work as a seq scan being performed.

**Comment by fritshoogland-yugabyte** [comment]
Based on the comments by @ramsrivatsa in the thread: 
A better understanding of what is happening can be seen when turning on `yb_debug_log_docdb_requests` at the YSQL level.

For the well performing case, the following operation can be seen:
```
I0628 13:01:55.342577  7855 pg_session.cc:258] Applying operation: { READ active: 1 read_time: { read: <invalid> local_limit: <invalid> global_limit: <invalid> in_txn_limit: <invalid> serial_no: 0 } request: client: YQL_CLIENT_PGSQL stmt_id: 78635104 schema_version: 0 targets { column_id: 0 } targets { column_id: 2 } targets { column_id: -8 } column_refs { ids: 0 ids: 2 } is_forward_scan: 1 is_aggregate: 0 limit: 1024 return_paging_state: 1 ysql_catalog_version: 242 table_id: "000033e5000030008000000000004a35" condition_expr { condition { op: QL_OP_AND operands { condition { op: QL_OP_BETWEEN operands { column_id: 0 } operands { value { int32_value: 39652 } } operands { value { int32_value: 39751 } } } } } } upper_bound { key: "4880009B477E21" is_inclusive: 1 } col_refs { column_id: 0 attno: 1 } col_refs { column_id: 2 attno: 3 } partition_key: "4880009AE421" }
```
Which does indicate the between operation is sent to DocDB: `QL_OP_BETWEEN` with a single RPC.

For the bad performing case, the following operations can be seen:
```
I0628 13:03:59.228495  7855 pg_session.cc:258] Applying operation: { READ active: 1 read_time: { read: <invalid> local_limit: <invalid> global_limit: <invalid> in_txn_limit: <invalid> serial_no: 0 } request: client: YQL_CLIENT_PGSQL stmt_id: 82423904 schema_version: 0 targets { column_id: 0 } targets { column_id: 2 } targets { column_id: -8 } column_refs { ids: 0 ids: 2 } is_forward_scan: 1 is_aggregate: 0 limit: 1024 return_paging_state: 1 ysql_catalog_version: 242 table_id: "000033e5000030008000000000004a35" col_refs { column_id: 0 attno: 1 } col_refs { column_id: 2 attno: 3 } }
I0628 13:03:59.235716  7855 pg_session.cc:258] Applying operation: { READ active: 1 read_time: { read: { physical: 1656421439229394 } local_limit: { physical: 1656421439229394 } global_limit: { physical: 1656421439729397 } in_txn_limit: <max> serial_no: 0 } request: client: YQL_CLIENT_PGSQL stmt_id: 82423904 schema_version: 0 targets { column_id: 0 } targets { column_id: 2 } targets { column_id: -8 } column_refs { ids: 0 ids: 2 } is_forward_scan: 1 is_aggregate: 0 limit: 1024 paging_state { next_partition_key: "48800004012123800171BC4BADF22D804A" next_row_key: "48800004012123800171BC4BADF22D804A" total_num_rows_read: 1024 read_time { read_ht: 6784702215083597824 deprecated_max_of_read_time_and_local_limit_ht: 6784702215083597824 global_limit_ht: 6784702217131610112 in_txn_limit_ht: 18446744073709551615 local_limit_ht: 6784702215083597824 } } return_paging_state: 1 table_id: "000033e5000030008000000000004a35" col_refs { column_id: 0 attno: 1 } col_refs { column_id: 2 attno: 3 } partition_key: "48800004012123800171BC4BADF22D804A" }
... many more rows, where total_num_rows_read increments by 1024
```
Total number of rows = 50,000 / 1024 = 48.8.
This explains the 49 iters for `rocksdb_no_table_cache_iterators`, `handler_latency_yb_tserver_TabletServerService_Read`, `handler_latency_yb_tserver_PgClientService_Perform`.

**Comment by fritshoogland-yugabyte** [comment]
ps. the prepared statement is not needed to force this behaviour, an explicit cast in a normal SQL can achieve the same:
```
select c from performance where id > 30000::bigint and id < 30001::bigint;
```

What is remarkable is that an equality predicate will allow conversion to integer:
```
I0628 13:45:47.872854  7855 pg_session.cc:258] Applying operation: { READ active: 1 read_time: { read: <invalid> local_limit: <invalid> global_limit: <invalid> in_txn_limit: <invalid> serial_no: 0 } request: client: YQL_CLIENT_PGSQL stmt_id: 88010848 schema_version: 0 targets { column_id: 0 } targets { column_id: 2 } targets { column_id: -8 } column_refs { ids: 0 ids: 2 } is_forward_scan: 1 is_aggregate: 0 limit: 1024 return_paging_state: 1 ysql_catalog_version: 242 range_column_values { value { int32_value: 30000 } } table_id: "000033e5000030008000000000004a35" upper_bound { key: "488000753021" is_inclusive: 1 } col_refs { column_id: 0 attno: 1 } col_refs { column_id: 2 attno: 3 } partition_key: "488000753021" }
```

**Comment by ramsrivatsa** [comment]
Fixed this issue with the following commit https://github.com/yugabyte/yugabyte-db/commit/53a77712b5feefa6ff8b7d22984662fd51829575
